Sources/HummingbirdRouter/RouteController.swift

```diff
@@ -0,0 +1,106 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Hummingbird server framework project
+//
+// Copyright (c) 2023-2024 the Hummingbird authors
+// Licensed under Apache License v2.0
+//
+// See LICENSE.txt for license information
+// See hummingbird/CONTRIBUTORS.txt for the list of Hummingbird authors
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+//===----------------------------------------------------------------------===//
+
+import Hummingbird
+
+// MARK: - RouteController
+
+/// A type that defines a body of routes.
+///
+/// You can create custom controllers by declaring types that conform to the `RouteController`
+/// protocol. Implement the required ``RouteController/body`` computed property and provide a
+/// type for the controller's `Context`.
+///
+///     struct UsersController {
+///         typealias Context = BasicRouterRequestContext
+///
+///         var body: some RouterMiddleware<Context> {
+///             RouteGroup("/users") {
+///                 ...
+///             }
+///         }
+///     }
+///
+/// Assemble the controller's body by combining one or more components together to create
+/// a route. By nesting controllers within each other, you can compose large and complex
+/// routes from smaller more mangeable components.
+public protocol RouteController where Context == Body.Context {
+    associatedtype Context: RouterRequestContext
+    associatedtype Body: RouterMiddleware
+    @RouteBuilder<Context> var body: Body { get }
+}
+
+// MARK: - MiddlewareFixedTypeBuilder + RouteController
+
+/// Middleware stack result builder
+///
+/// Generates a middleware stack from the elements inside the result builder. The input,
+/// context and output types passed through the middleware stack are fixed and cannot be changed.
+extension MiddlewareFixedTypeBuilder {
+    public static func buildExpression<C0: RouteController>(_ c0: C0) -> C0.Body where C0.Body.Input == Input, C0.Body.Output == Output, C0.Body.Context == Context {
+        return c0.body
+    }
+
+    public static func buildBlock<C0: RouteController>(_ c0: C0) -> C0.Body {
+        return c0.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController>(first: C0) -> C0.Body {
+        first.body
+    }
+
+    public static func buildPartialBlock<C0: RouteController, C1: RouteController>(
+        accumulated c0: C0,
+        next c1: C1
+    ) -> _Middleware2<C0.Body, C1.Body> where C0.Body.Input == C1.Body.Input, C0.Body.Output == C1.Body.Output, C0.Body.Context == C1.Body.Context {
+        _Middleware2(c0.body, c1.body)
+    }
+
+    public static func buildPartialBlock<C0: RouteController, M0: MiddlewareProtocol>(
+        accumulated c0: C0,
+        next m0: M0
+    ) -> _Middleware2<C0.Body, M0> where C0.Body.Input == M0.Input, C0.Body.Output == M0.Output, C0.Body.Context == M0.Context {
```

Comment:
The accumulated middleware should only ever be a MiddlewareProtocol so I don't think this function is necessary.